crates/tedge_cli/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+        #[structopt(long = "all")]
+        is_all: bool,
+
+        /// Prints all keys and descriptions with example values
+        #[structopt(long = "doc")]
+        is_doc: bool,
+    },
+}
+
+#[derive(StructOpt, Debug)]
+enum ConnectCommandCli {
+    // ...
+}
+
+#[derive(StructOpt, Debug)]
+enum DisconnectCommandCli {
+    // ...
+}
+
+
+// This would ideally live in `tedge_command_tedge`.
+pub enum TEdgeCommand {
+    CertCommand(tedge_command_cert::CertCommand),
+    GetConfigCommand(tedge_command_config::GetConfigCommand),
+    ConnectCommand(tedge_command_connect::ConnectCommand),
+    // ...
+}
+
+impl Command for TEdgeCommand {
+    type Error = anyhow::Error; // .. either the top-level error type that accumulates all the sub-command error types, or just anyhow::Error
+
+    fn description(&self) -> String {
+        match self {
+            Self::CertCommand(cmd) => cmd.description(),
+            Self::GetConfigCommand(cmd) => cmd.description(),
+            Self::ConnectCommand(cmd) => cmd.description(),
+        }
+    }
+
+    fn execute(self) -> Result<(), Self::Error> {
+        match self {
+            Self::CertCommand(cmd) => cmd.execute(),
+            Self::GetConfigCommand(cmd) => cmd.execute(),
+            Self::ConnectCommand(cmd) => cmd.execute(),
+        }
+    }
+}
+ 
+impl TEdgeCli {
+    pub fn into_command(self, context: ...) -> Result<TEdgeCommand, CliError> {
+        match self.tedge_command {
+            TEdgeCommandCli::Cert(opt) => {
+                unimplemented!()
+                // ...
+            }
+            TEdgeCommandCli::Config(config_cmd) => {
+                match config_cmd {
+                    ConfigCommandCli::Get { key } => Ok(
+                        TEdgeCommand::GetConfigCommand(
+                        tedge_command_config::GetConfigCommand {
+                            config_key: key,
+                            config
+                        }))
+
+                    // ...
+                }
+            }
+            TEdgeOpt::Connect(opt) => {
+                unimplemented!()
+            }
+            TEdgeOpt::Disconnect(opt) => {
+                unimplemented!()
+            }
+            TEdgeOpt::Mqtt(opt) => {
+                unimplemented!()
+            }
+        }
+
+    }
+}
+
+// These kind of tests ensure that our CLI parser works correctly.
+#[test]
+fn test_cli_parsing() {
+    let cli: TEdgeCli = TEdgeCli::parse_from("tedge config set device.id 123")?; 
+    assert_matches!(cli, TEdgeCli {
+        tedge_command: TEdgeCommandCli::Config(
+                           ConfigCommandCli::Set {
+                               key: tedge_command_config::ConfigKey::from_str("device.id")?,
+                               value: "123",
+                           }
+        )
+    });
+}
+
+// These kind of tests ensure that given a TEdgeCli, we convert it into the correct commands.
```

Comment:
Tests the TEdgeCli to TEdgeCommand mapping. Both are data structures that we can simply pattern match. There should not be any side-effects, so pretty simple to write unit tests.